tsdb/docs/usage.md

```diff
@@ -0,0 +1,152 @@
+
+
+## Adding data
+
+Use [`db.Appender()`](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#DB.Appender) to obtain an "appender".
+The [golang docs](https://pkg.go.dev/github.com/prometheus/prometheus/storage#Appender) speak mostly for themselves.
+
+Remember:
+
+* Use Commit() to add the samples to the DB and update the WAL.
+* create a new appender each time you commit.
+* appenders are not concurrency safe, but scrapes run concurrently and as such, leverage multiple appenders concurrently.
+  This reduces contention, although Commit() contend the same critical section (writing to the WAL is serialized), and may
+  inflate append tail latency if multiple appenders try to commit at the same time.
+
+Append may reject data due to these conditions:
+
+1) `timestamp < minValidTime` where `minValidTime` is the highest of:
+  * the maxTime of the last block (i.e. the last truncation time of Head) - updated via [`Head.Truncate()`](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#Head.Truncate) and [`DB.CompactHead()`](https://github.com/prometheus/prometheus/blob/1270b87970baeb926fcce64552db5c744ffaf83f/tsdb/db.go#L968)
+  * `tsdb.min-block-duration/2` older than the max time in the Head chunk. Note that while technically `storage.tsdb.min-block-duration` is configurable, it's a hidden option and changing it is discouraged.  So We can assume this value to be 2h.
+  Breaching this condition results in "out of bounds" errors.  
+  The first condition assures the block that will be generated doesn't overlap with the previous one (which simplifies querying)  
+  The second condition assures the sample won't go into the so called "compaction window", that is the section of the data that might be in process of being saved into a persistent block on disk.  (because that logic runs concurrently with ingestion without a lock)
+2) The labels don't validate. (if the set is empty or contains duplicate label names)
+3) If the sample, for the respective series (based on all the labels) is out of order or has a different value for the last (highest) timestamp seen. (results in storage.ErrOutOfOrderSample or storage.ErrDuplicateSampleForTimestamp respectively)
+
+Commit() may also refuse data that is out of order with respect to samples that were added via a different appender.
+
+Example:
+
+```go
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/tsdb"
+)
+
+func noErr(err error) {
+	if err != nil {
+		panic(err)
+	}
+}
+
+func main() {
+	db, err := tsdb.Open("tsdb-test", nil, nil, tsdb.DefaultOptions(), nil)
+	noErr(err)
+
+	ctx := context.Background()
+	app := db.Appender(ctx)
+
+	series := labels.FromStrings("foo", "bar")
+	ref, err := app.Append(0, series, time.Now().Unix(), 123)
+	time.Sleep(time.Second)
+	_, err = app.Append(ref, series, time.Now().Unix(), 124)
+	err = app.Commit()
+	noErr(err)
+	app = db.Appender(ctx) // need a new one. See app.Commit() docs
```

Comment:
Added comments and minor edits to existing comments.

```suggestion
	series := labels.FromStrings("foo", "bar")
	// ref is 0 for the first append since we don't know the reference for the series.
	ref, err := app.Append(0, series, time.Now().Unix(), 123)
	time.Sleep(time.Second)
	// Re-using the ref from above since it's the same series, makes append faster.
	_, err = app.Append(ref, series, time.Now().Unix(), 124)
	err = app.Commit()
	noErr(err)
	app = db.Appender(ctx) // Need a new one. See app.Commit() docs.
```